File: 772 Basic Calculator III.py
'''
772. Basic Calculator III
https://leetcode.com/problems/basic-calculator-iii/

Implement a basic calculator to evaluate a simple expression string.

The expression string contains only non-negative integers, '+', '-', '*', '/' operators,
and open '(' and closing parentheses ')'.
The integer division should truncate toward zero.

You may assume that the given expression is always valid.
All intermediate results will be in the range of [-231, 231 - 1].

Note: You are not allowed to use any built-in function which evaluates strings as mathematical expressions, such as eval().

Example 1:

Input: s = "1+1"
Output: 2

Example 2:

Input: s = "6-4/2"
Output: 4

Example 3:

Input: s = "2*(5+5*2)/3+(6/2+8)"
Output: 21

Constraints:
    1 <= s <= 104
    s consists of digits, '+', '-', '*', '/', '(', and ')'.
    s is a valid expression.
'''
class Solution:
    def calculate(self, s: str) -> int:
        stack = [] # [operator or number]
        number = 0
        operator = '+' # can be '+', '-', '*', '/',  or None

        # Compute whenever number parsing is done. 3 possibilities:
        # 1. Meet an operator
        # 2. Meet ')'
        # 3. End of s.
        def compute():
            nonlocal stack, number, operator
            if operator == '+':
                stack.append(number)
            elif operator == '-':
                stack.append(-number)
            elif operator == '*':
                stack[-1] = stack[-1] * number
            elif operator == '/':
                result = stack[-1] // number
                if result < 0 and stack[-1] % number != 0:
                    # This question asked to truncate toward zero
                    result += 1 
                stack[-1] = result
            # operator can be None too after parsing ')'
            
        for c in s:
            if c.isdigit():
                number = number * 10 + int(c)
            elif c in ['+', '-', '*', '/']:
                compute()
                operator = c
                number = 0
            elif c == '(':
                # There must be an operator ahead of '('
                stack.append(operator)
                operator = '+'
                # number = 0 # Not needed as number is set to 0 on previous operator
            elif c == ')':
                compute()

                # Get sum of the numbers inside parentheses,
                # compute with the operator in front of '('
                number = 0
                while isinstance(stack[-1], int):
                    number += stack.pop()
                operator = stack.pop()
                compute()

                # As next char must be an operator,
                # set to None to avoid duplicate computing
                operator = None 
                number = 0

        compute()
        return sum(stack)

File: 772 Basic Calculator III.ts
/*
772. Basic Calculator III
https://leetcode.com/problems/basic-calculator-iii
Consider '+', '-', '*', '/', '(', ')'
*/
type StackMember = number | string;

function compute(num: number, op: string | undefined, stack: StackMember[]) {
    // For '*', '/', compute current num with previous num.
    // For '+', '-', not compute but push the value to the stack,
    // in case there is a '*', '/' following up.
    if (op === '+') {
        stack.push(num);
    } else if (op === '-') {
        stack.push(-num);
    } else if (op === '*') {
        stack.push((stack.pop() as number) * num);
    } else if (op === '/') {
        console.log(op, stack, num);
        const r = (stack.pop() as number) / num;
        if (r > 0) {
            stack.push(Math.floor(r));
        } else {
            stack.push(Math.ceil(r));
        }
    }
}

function collapse(stack: StackMember[]) {
    // Collapse the stack to get a single number,
    // then compute with previous operator and number
    let num = 0;
    while (typeof stack[stack.length - 1] === 'number') {
        num += stack.pop() as number;
    }
    // There must be an operator ahead
    const op = stack.pop() as string;
    compute(num, op, stack);
}

function calculate(s: string) {
    let op: string | undefined = '+';

    // A stack that saves previous numbers,
    // in case of '(', it saves previous operators
    const stack: StackMember[] = [];
    // Represents current number in parsing
    let num = 0;

    for (const char of s) {
        if (['+', '-', '*', '/'].includes(char)) {
            compute(num, op, stack);
            op = char;
            num = 0;
        } else if (char === '(') {
            stack.push(op!);
            op = '+';
            num = 0;
        } else if (char === ')') {
            compute(num, op, stack);
            collapse(stack);
            op = undefined; // must be undefined, as next char should be the operator
            num = 0;
        } else {
            const digit = parseInt(char);
            if (!isNaN(digit)) {
                num = num * 10 + digit;
            }
        }
    }
    compute(num, op, stack);
    return (stack as number[]).reduce((prev, curr) => prev + curr, 0);
}
